### PR TITLE
[NOD-1001] Make an error in Peer.start() stop the connection process from continuing.

### PR DESCRIPTION
**Pull request by svarogg**

### CODE REVIEW

File: peer/example_test.go [hunk 1]
```diff
@@ -38,23 +39,29 @@ func mockRemotePeer() error {
...
 
 	return nil
 }
 
 // This example demonstrates the basic process for initializing and creating an
-// outbound peer. Peers negotiate by exchanging version and verack messages.
+// outbound  Peers negotiate by exchanging version and verack messages.
 // For demonstration, a simple handler for version message is attached to the
-// peer.
+//
```
Comment (stasatdaglabs) [inline peer/example_test.go:57]: The above changes seem to be a `replace` error.

File: peer/example_test.go [hunk 2]
```diff
@@ -38,23 +39,29 @@ func mockRemotePeer() error {
...
 func Example_newOutboundPeer() {
+	log.SetLevel(logs.LevelOff) // Disable logs to not interfer with output
```
Comment (stasatdaglabs) [inline peer/example_test.go:59]: Typo: interfer -> interfere

File: peer/example_test.go [hunk 3]
```diff
@@ -38,23 +39,29 @@ func mockRemotePeer() error {
...
+
 	// Ordinarily this will not be needed since the outbound peer will be
 	// connecting to a remote peer, however, since this example is executed
 	// and tested, a mock remote peer is needed to listen for the outbound
-	// peer.
+	//
```
Comment (stasatdaglabs) [inline peer/example_test.go:64]: The above change seem to be a `replace` error.

File: peer/example_test.go [hunk 4]
```diff
@@ -101,7 +112,7 @@ func Example_newOutboundPeer() {
 		fmt.Printf("Example_peerConnection: verack timeout")
 	}
 
-	// Disconnect the peer.
+	// Disconnect the
```
Comment (stasatdaglabs) [inline peer/example_test.go:115]: The above change seem to be a `replace` error.

File: peer/peer_test.go [hunk 5]
```diff
@@ -93,7 +95,7 @@ func pipe(c1, c2 *conn) (*conn, *conn) {
 	return c1, c2
 }
 
-// peerStats holds the expected peer stats used for testing peer.
+// peerStats holds the expected peer stats used for testing
```
Comment (stasatdaglabs) [inline peer/peer_test.go:98]: The above change seem to be a `replace` error.

File: peer/peer_test.go [hunk 6]
```diff
@@ -531,18 +497,16 @@ func TestOutboundPeer(t *testing.T) {
...
 
-	r, w := io.Pipe()
-	c := &conn{raddr: "10.0.0.1:16111", Writer: w, Reader: r}
-
-	p, err := peer.NewOutboundPeer(peerCfg, "10.0.0.1:16111")
+	_, p, err := setupPeers(peerCfg, peerCfg)
 	if err != nil {
-		t.Errorf("NewOutboundPeer: unexpected err - %v\n", err)
-		return
+		t.Fatalf("TestOuboundPeer: unexpected in err setupPeers - %v\n", err)
```
Comment (stasatdaglabs) [inline peer/peer_test.go:502]: Typo: "in err" -> "err in"

File: peer/peer_test.go [hunk 7]
```diff
@@ -572,14 +536,11 @@ func TestOutboundPeer(t *testing.T) {
...
 	peerCfg.SelectedTipHash = selectedTipHash
-	r1, w1 := io.Pipe()
-	c1 := &conn{raddr: "10.0.0.1:16111", Writer: w1, Reader: r1}
-	p1, err := peer.NewOutboundPeer(peerCfg, "10.0.0.1:16111")
+
+	_, p1, err := setupPeers(peerCfg, peerCfg)
 	if err != nil {
-		t.Errorf("NewOutboundPeer: unexpected err - %v\n", err)
-		return
+		t.Fatalf("TestOuboundPeer: unexpected in err setupPeers - %v\n", err)
```
Comment (stasatdaglabs) [inline peer/peer_test.go:542]: Typo: "in err" -> "err in"

File: peer/example_test.go (hunk 1 above)
Comment (svarogg) [reply peer/example_test.go:57]: LOL, yeah, I moved the whole thing out of peer_test, so had to remove all `peer.`

File: peer/peer_test.go [hunk 8]
```diff
@@ -200,12 +202,12 @@ func testPeer(t *testing.T, p *peer.Peer, s peerStats) {
 // TestPeerConnection tests connection between inbound and outbound peers.
 func TestPeerConnection(t *testing.T) {
 	verack := make(chan struct{})
```
Comment (someone235) [inline peer/peer_test.go:204]: I would prefer that each peer will have its own channels, separate for input and output

File: peer/peer_test.go [hunk 9]
```diff
@@ -730,9 +697,56 @@ func TestUnsupportedVersionPeer(t *testing.T) {
...
+func setupPeersWithConns(inPeerCfg, outPeerCfg *Config, inConn, outConn *conn) (inPeer *Peer, outPeer *Peer, err error) {
+	inPeer = NewInboundPeer(inPeerCfg)
+	inPeerDone := make(chan struct{})
+	var inPeerErr error
+	go func() {
+		inPeerErr = inPeer.AssociateConnection(inConn)
+		inPeerDone <- struct{}{}
+	}()
+
+	outPeer, err = NewOutboundPeer(outPeerCfg, "10.0.0.2:16111")
```
Comment (someone235) [inline peer/peer_test.go:724]: Shouldn't it be an argument or at least a constant?

File: peer/peer_test.go [hunk 10]
```diff
@@ -730,9 +697,56 @@ func TestUnsupportedVersionPeer(t *testing.T) {
...
+		outPeerDone <- struct{}{}
+	}()
+
+	// wait for AssociateConnection to complete in all instances
+	if !testtools.WaitTillAllCompleteOrTimeout(2*time.Second, inPeerDone, outPeerDone) {
+		return nil, nil, errors.New("handshake timeout")
+	}
+
+	if inPeerErr != nil && outPeerErr != nil {
+		return nil, nil, errors.Errorf("Both inPeer and outPeer failed connecting: \nInPeer: %+v\nOutPeer: %+v",
```
Comment (someone235) [inline peer/peer_test.go:741]: Errors should begin with lower case

File: peer/peer_test.go (hunk 8 above)
Comment (svarogg) [reply peer/peer_test.go:204]: Why? These are just to make sure that handshake completed

File: peer/peer_test.go [hunk 11]
```diff
@@ -40,6 +40,11 @@ type conn struct {
 	proxy bool
 }
 
+const (
+	addr1 = "10.0.0.1:16111"
```
Comment (someone235) [inline peer/peer_test.go:44]: Please use more meaning names, like outPeerAddr or something like this

File: peer/peer_test.go (hunk 8 above)
Comment (someone235) [reply peer/peer_test.go:209]: After a discussion we decided to fix this

File: peer/peer_test.go (hunk 11 above)
Comment (svarogg) [reply peer/peer_test.go:44]: It really depends on the test.
The addresses have no meaning, just something that you need to specify to make things work.

**Comment by svarogg** [comment]
Rebuild after update in master

File: peer/example_test.go [hunk 12]
```diff
@@ -38,11 +45,16 @@ func mockRemotePeer() error {
 			return
 		}
 
-		// Create and start the inbound peer.
+		// Create and start the inbound
```
Comment (stasatdaglabs) [inline peer/example_test.go:48]: This appears to be a `replace` error.

File: peer/example_test.go (hunk 3 above)
Comment (stasatdaglabs) [reply peer/example_test.go:70]: Please remove the superfluous `//`.

File: peer/example_test.go [hunk 13]
```diff
@@ -38,11 +45,16 @@ func mockRemotePeer() error {
...
 		p := peer.NewInboundPeer(peerCfg)
-		p.AssociateConnection(conn)
+		err = p.AssociateConnection(conn)
+		if err != nil {
+			fmt.Printf("AssociateConnection: error %+v\n", err)
+			return
+		}
 	}()
 
+	<-time.After(1 * time.Second)
```
Comment (stasatdaglabs) [inline peer/example_test.go:57]: I'm not sure I'm a fan of this. Every time a test calls this function the whole test runtime becomes 1 second slower.

Comment (svarogg) [reply peer/example_test.go:57]: Oh yeah, it's a terrible idea in all 100%. Added it when checking something and forgot to remove.

Good catch!

**Comment by svarogg** [comment]
Rebuild after upgrade in master

**Comment by svarogg** [comment]
Rebuild after update in master

**Pull request merged**